File: 0047-permutations-ii/0047-permutations-ii.py
class Solution(object):
    def permuteUnique(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        N = len(nums)
        used = [0]*N

        def backtrack_permute(arr, result):
            if len(arr) == N and arr not in result :
                result.append(arr[:])
                return 

            for i in range(N) :
                if not used[i] :
                    arr.append(nums[i])
                    used[i] = 1
                    backtrack_permute(arr, result)
                    used[i] = 0
                    arr.pop()
        
        nums.sort()
        result = []
        backtrack_permute([],result)
        return result

File: 0047-permutations-ii/0047-permutations-ii.c
class Solution {
public:
    void backtrackUnique(vector<int>& arr, vector<int>& used, vector<vector<int>>& result, vector<int>& nums, int start){
        if (arr.size() == nums.size()) {
            result.push_back(arr);
            return;
        }
        for (int i = start; i < nums.size(); i++) {
            if (i > 0 && nums[i] == nums[i - 1] && !used[i - 1]) continue;

            if (!used[i]) {
                used[i] = 1;
                arr.push_back(nums[i]);
                backtrackUnique(arr, used, result, nums, 0);
                used[i] = 0;
                arr.pop_back();
            }
        }
    }

    vector<vector<int>> permuteUnique(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<int> arr;
        vector<int> used(nums.size(), 0);
        vector<vector<int>> result;

        backtrackUnique(arr, used, result, nums, 0); 
        return result;        
    }
};
